generators/Program.cs

```diff
@@ -35,30 +35,69 @@ private static void SetupLogger()
 
         private static void RegenerateTestClasses(Options options)
         {
+            options.Normalize();
+
+            var canonicalDataFile = new CanonicalDataFile(options);
+            canonicalDataFile.DownloadData();
+
             Log.Information("Re-generating test classes...");
             
-            var canonicalDataParser = CreateCanonicalDataParser(options);
+            var canonicalDataParser = new CanonicalDataParser(canonicalDataFile);
 
-            foreach (var exercise in new ExerciseCollection(options.Exercises))
-            {
-                var canonicalData = canonicalDataParser.Parse(exercise);
-                exercise.Regenerate(canonicalData);
-            }
+            foreach (var exercise in new ExerciseCollection(canonicalDataFile))
+                RegenerateTestClass(exercise, options, canonicalDataParser);
 
             Log.Information("Re-generated test classes.");
         }
 
-        private static CanonicalDataParser CreateCanonicalDataParser(Options options)
+        private static void RegenerateTestClass(Exercise exercise, Options options, CanonicalDataParser canonicalDataParser)
         {
-            var canonicalDataOptions = new CanonicalDataOptions
+            if (ShouldBeSkipped(exercise, options))
+                return;
+
+            switch (exercise)
             {
-                CanonicalDataDirectory = options.CanonicalDataDirectory ?? DefaultCanonicalDataDirectory,
-                CacheCanonicalData = options.CacheCanonicalData
-            };
-            return new CanonicalDataParser(canonicalDataOptions);
+                case GeneratorExercise generatorExercise:
+                    var canonicalData = canonicalDataParser.Parse(exercise.Name);
+                    generatorExercise.Regenerate(canonicalData);
+
+                    Log.Information("{Exercise}: tests generated", exercise.Name);
+                    break;
+                case UnimplementedExercise _:
```

Comment:
Can't you remove the underscores here since they're not used?

Reply:
Nope, unfortunately not. I tried that at first, but then it isn't a valid pattern match anymore.

Reply:
Hmm ok I thought you could, thanks for checking!